src/render/painter.js

```diff
@@ -344,48 +347,77 @@ class Painter {
             return;
         }
 
+        const renderableSkippedTileIDs = [];
         if (this._tileClippingMaskIDs && !this.terrain) {
             let dirtyStencilClippingMasks = false;
             // Equivalent tile set is already rendered in stencil
             for (const coord of tileIDs) {
                 if (this._tileClippingMaskIDs[coord.key] === undefined) {
                     dirtyStencilClippingMasks = true;
-                    break;
+                }
+                if (this._skippedStencilTileIDs.includes(coord.key)) {
```

Comment:
What's the typical length of this array on average? If it can be relatively long, should we consider other data structures like a `Set` to avoid quadratic behavior?

Reply:
Hmm I think normally we don't have many tiles at the same time in a frame, but maybe there are some corner cases with a large viewport and with a lot of child tiles where we can have a couple dozens. But I think using a set is a good idea in any case.